compiler/rustc_lint/src/types.rs

```diff
@@ -884,9 +884,33 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
             if let Some(field) = transparent_newtype_field(self.cx.tcx, variant) {
                 self.check_field_type_for_ffi(cache, field, substs)
             } else {
-                // All fields are ZSTs; this means that the type should behave
-                // like (), which is FFI-unsafe
+                // All fields are ZSTs; this means that except if all
+                // fields are PhantomData the type should behave like (),
+                // which is FFI-unsafe,
+                let mut all_phantom = !variant.fields.is_empty();
+                for field in &variant.fields {
```

Comment:
Could you attempt to deduplicate this a little with the logic in the `else` block that immediately follows this one?

Reply:
Good question. This is what I tried originally, but I have always ended up with more akward code that what is here. A notable issue is to output the correct error message. If you delay the emission of `lint_improper_ctypes_struct_zst` until you have gone through the field loop, you end up emitting the wrong error message because of the `r => r` match arm, and trying to remove it led to awkward code iirc.

Reply:
It might be okay if another lint was emitted first before `lint_improper_ctypes_struct_zst` - that might not be what we have in the output now, but it would still be correct, I think (because if you fixed that lint, you would then get the `lint_improper_ctypes_struct_zst` after).

Reply:
What about something like this?

```rust
let maybe_default_result = if def.repr().transparent() {
    if let Some(field) = transparent_newtype_field(self.cx.tcx, variant) {
        self.check_field_type_for_ffi(cache, field, substs)
    } else {
        Some(FfiUnsafe { ty, reason: fluent::lint_improper_ctypes_struct_zst, help: None })
    }
} else {
    None
};

let mut all_phantom = !variant.fields.is_empty();
for field in &variant.fields {
    match self.check_field_type_for_ffi(cache, &field, substs) {
        /* what you have */
        r => return maybe_default_result.unwrap_or(r),
    }
}

if all_phantom { FfiPhantom(ty) } else { maybe_default_result.unwrap_or(FfiSafe) }
```

Reply:
This seems to work fine and to work on my example. Thanks! :)